numpy/lib/_function_base_impl.py

```diff
@@ -4535,8 +4535,26 @@ def quantile(a,
         if axis is not None:
             axis = _nx.normalize_axis_tuple(axis, a.ndim, argname="axis")
         weights = _weights_are_valid(weights=weights, a=a, axis=axis)
+        if weights.dtype != object:
```

Comment:
A general comment: I think these checks should happen inside `_weights_ar_valid` - this will ensure they are used for `percentile` as well.